### PR TITLE
Louis/bug-fixes

### PR DESCRIPTION
**Pull request by louislefevre**
### Purpose
Some bug fixes for issues #55 and #59, and disabled flaky tests.

### Changes

- Fix strings which were being formatted incorrectly (583f6cbb61c080d1ccc728b79ea29036a6e85229).
- Fix qubits not resetting on error (b70d6bebf7ae932c39f2ee581d45b39e0f31f21c).
- Disabled tests for errors until issue #58 is resolved (dba788a2e8694119f0d39da18dbf9c0a96ba0bf7).

### CODE REVIEW

File: quantum/nodes/quantum-register/quantum-register.js [hunk 1]
```diff
@@ -97,13 +97,16 @@ module.exports = function(RED) {
...
       // then send one qubit object per node output
-      await shell.execute(script, (err) => {
-        if (err) done(err);
+      await shell.execute(script, async (err) => {
+        if (err) {
+          done(err);
+        } else {
+          // wait for quantum circuit to be initialised
+          await circuitReady();
+          send(output);
```
Comment (Theo-Reignier) [inline quantum/nodes/quantum-register/quantum-register.js:106]: I believe runtime variables should be emptied here as well

Comment (louislefevre) [reply quantum/nodes/quantum-register/quantum-register.js:106]: As in we should reset `output` to an empty array?

**Review by Theo-Reignier — CHANGES_REQUESTED** [review]
It would be cleaner to also empty the `output` runtime variable for the quantum register node.

Finally, some nodes throw errors during their execution and before `shell.execute`.
If these nodes have runtime variables, then they should empty them at the same time.

File: quantum/nodes/quantum-register/quantum-register.js (hunk 1 above)
Comment (Theo-Reignier) [reply quantum/nodes/quantum-register/quantum-register.js:106]: You are right, it does not change anything.
Would be cleaner though.

**Comment by louislefevre** [comment]
> It would be cleaner to also empty the `output` runtime variable for the quantum register node.

Some other nodes, such as the barrier node, also have an `output` variable. Shall I reset these as well?

> Finally, some nodes throw errors during their execution and before `shell.execute`.
> If these nodes have runtime variables, then they should empty them at the same time.

Good point, I'll get on that 👍

**Comment by Theo-Reignier** [comment]
It wouldn't cause an issue if you don't reset all `output` variables but I believe it would be cleaner to empty all of them.

**Comment by louislefevre** [comment]
@Theo-Reignier Alternatively, do you think it would be simpler to initialise the variables at the top of the input listener?

Right now, variables are being initialised when the node is created.
```node
function BarrierNode(config) {
    RED.nodes.createNode(this, config);
    this.name = config.name;
    this.outputs = parseInt(config.outputs);
    this.qubits = [];
    const node = this;

    this.on('input', async function(msg, send, done) {
      ...
    });
```

Could we move this initialisation to the inside of the input listener like so:
```node
function BarrierNode(config) {
    RED.nodes.createNode(this, config);
    this.name = config.name;
    const node = this;

    this.on('input', async function(msg, send, done) {
      node.qubits = [];
      node.outputs  = parseInt(config.outputs);
      ...
    });
```

This would ensure that they are reset every time they receive new input.

**Comment by Theo-Reignier** [comment]
That would simplify the code indeed. Good idea !

However, I believe that all the initializations based on `config.xxx` can stay before the input listener.
I believe it makes more sense and is more efficient

**Comment by louislefevre** [comment]
> However, I believe that all the initializations based on `config.xxx` can stay before the input listener.

Makes sense, given that the value of `outputs` will be the same regardless of the input being received by the listener anyway, and it isn't being modified anywhere. In that case, in combination with my previous idea, I think leaving `outputs` as it is without resetting it would be fine.

**Comment by Theo-Reignier** [comment]
Okay then, I am happy with that.

**Comment by louislefevre** [comment]
Oh I misunderstood - you was referring to `output`, but I thought you meant `outputs`. I'll stick to resetting `output` in that case 
👍

**Comment by Theo-Reignier** [comment]
I misunderstood your message as well, sorry about that !

**Comment by louislefevre** [comment]
I realise I don't think my idea of putting `node.qubits = [];` at the top of the input listener will work, because sometimes we don't want it to reset on each input. The simulator node for example, we don't want to reset `qubits` on every input because we need to push every input `msg` we get to the same array, so  I'll go back to the original method of resetting.

**Comment by Theo-Reignier** [comment]
That's true, I didn't realise that while discussing here.

**Comment by louislefevre** [comment]
@Theo-Reignier I've gone with a method of creating `reset()` functions and calling them after each `done()`. What do you think? (1e3a4f7f393b6ca640d954b3f40cf9b27c0256c7)

**Comment by Theo-Reignier** [comment]
Much much better solution, way cleaner as well. 
Amazing idea, thank Louis !

Is that ready for review ?

**Comment by louislefevre** [comment]
> Much much better solution, way cleaner as well.
> Amazing idea, thank Louis !
> 
> Is that ready for review ?

Thanks! 😊
Yup it's ready for review.

**Review by Theo-Reignier — APPROVED** [review]
All good

**Comment by louislefevre** [comment]
@bbaayezi Can you have a look at 583f6cbb61c080d1ccc728b79ea29036a6e85229 for issue #55? I think I found all the string formatting errors.

**Comment by bbaayezi** [comment]
> @bbaayezi Can you have a look at [583f6cb](https://github.com/louislefevre/node-red-contrib-quantum/commit/583f6cbb61c080d1ccc728b79ea29036a6e85229) for issue #55 please? I think I found all the string formatting errors.

I think you stil missed the `identity-gate`, other than that LGTM.

**Comment by louislefevre** [comment]
> I think you stil missed the `identity-gate`, other than that LGTM.

[It seems to be okay I think](https://github.com/louislefevre/node-red-contrib-quantum/blob/dba788a2e8694119f0d39da18dbf9c0a96ba0bf7/quantum/nodes/identity-gate/identity-gate.js#L29)? Or did you mean I should make them consistent in format (e.g. update this one to be  `script += util.format(snippets.IDENTITY, msg.payload.registerVar + '[' + msg.payload.qubit + ']');`)?

**Comment by bbaayezi** [comment]
> > I think you stil missed the `identity-gate`, other than that LGTM.
> 
> [It seems to be okay I think](https://github.com/louislefevre/node-red-contrib-quantum/blob/dba788a2e8694119f0d39da18dbf9c0a96ba0bf7/quantum/nodes/identity-gate/identity-gate.js#L29)? Or did you mean I should make them consistent in format (e.g. update this one to be `script += util.format(snippets.IDENTITY, msg.payload.registerVar + '[' + msg.payload.qubit + ']');`)?

ah okay I misread it then.

File: test/python.test.js [hunk 2]
```diff
@@ -163,7 +163,7 @@ describe('PythonShell', function() {
       assert.strictEqual(output, '0\n1\n2\n3');
     });
 
-    it('return error on invalid command', async function() {
+    xit('return error on invalid command', async function() {
```
Comment (bbaayezi) [inline test/python.test.js:166]: does the `xit()` function means skipping the test for now?

Comment (louislefevre) [reply test/python.test.js:166]: Yeah - these tests are flaky due to issue #58 and were causing builds to fail arbitrarily, so I've disabled them for now.

**Comment by louislefevre** [comment]
@bbaayezi Could you approve this PR if you're ready please?

**Pull request merged**